File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def myFun(level,arr):
            if level == len(arr):
                res.append(arr[:])
            for i in range(level,len(nums)):
                new_arr = arr[:]
                new_arr[i],new_arr[level] = new_arr[level],new_arr[i]
                myFun(level + 1,new_arr)
        myFun(0,nums)
        return res
            
                    

File: 46-permutations/permutations.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    int[] numsG;

    public List<List<Integer>> permute(int[] nums) {
        numsG = nums;
        boolean[] used = new boolean[nums.length];
        backtrack(new ArrayList<>(), used);
        return result;
    }

    private void backtrack(List<Integer> curr, boolean[] used) {
        if (curr.size() == numsG.length) {
            result.add(new ArrayList<>(curr));
            return;
        }

        for (int i = 0; i < numsG.length; i++) {
            if (used[i]) continue;      // already used

            used[i] = true;
            curr.add(numsG[i]);

            backtrack(curr, used);

            curr.remove(curr.size() - 1);
            used[i] = false;
        }
    }
}
